File: Shared/Engine/DayStamp.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import Foundation

/// A calendar day as an integer — days since 1970-01-01 (proleptic Gregorian).
///
/// A training day is the day the user lived through: a session finished at 00:30
/// belongs to the evening before (the day turns at `rolloverHour`), and "2 of 2 today"
/// must flip at that hour without a relaunch. Integers make that pure arithmetic,
/// immune to the off-by-one a local-midnight `Date` re-read in another zone invites,
/// and make `WorkoutLog.dayKey` a cheap Int predicate.
///
/// The ONLY place `Calendar` appears is this file's conversion boundary: take Y/M/D in
/// the user's calendar, rebuild it in a fixed UTC Gregorian calendar, divide by 86 400.
struct DayStamp: Hashable, Comparable, Sendable, Strideable {
    var raw: Int

    init(raw: Int) { self.raw = raw }

    static func < (lhs: DayStamp, rhs: DayStamp) -> Bool { lhs.raw < rhs.raw }

    // Strideable — lets ranges of days iterate naturally (the 14-day consistency strip
    // is literally `(today - 13)...today`).
    func advanced(by n: Int) -> DayStamp { DayStamp(raw: raw + n) }
    func distance(to other: DayStamp) -> Int { other.raw - raw }

    static func + (lhs: DayStamp, rhs: Int) -> DayStamp { DayStamp(raw: lhs.raw + rhs) }
    static func - (lhs: DayStamp, rhs: Int) -> DayStamp { DayStamp(raw: lhs.raw - rhs) }
    /// Signed distance in days (NOT a count of inclusive days — that is this + 1).
    static func - (lhs: DayStamp, rhs: DayStamp) -> Int { lhs.raw - rhs.raw }
}

// MARK: - Codable (a bare Int on the wire)

extension DayStamp: Codable {
    init(from decoder: Decoder) throws {
        raw = try decoder.singleValueContainer().decode(Int.self)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        try c.encode(raw)
    }
}

// MARK: - Date boundary (the only Calendar use in the engine)

extension DayStamp {
    /// Fixed UTC Gregorian calendar: the neutral frame every Y/M/D is rebuilt in, so
    /// `raw` is exact division and never affected by the device's zone or DST.
    static let utcCalendar: Calendar = {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = TimeZone(identifier: "UTC")!
        return c
    }()

    private static let secondsPerDay: Double = 86_400

    /// A Gregorian calendar in the given calendar's TIME ZONE — never its calendar
    /// system: Buddhist or Japanese years (2569, Reiwa 8) rebuilt under Gregorian rules
    /// would shift every stamp by centuries.
    static func gregorian(zoneOf calendar: Calendar) -> Calendar {
        var g = Calendar(identifier: .gregorian)
        g.timeZone = calendar.timeZone
        return g
    }

    /// The calendar day `date` falls on in `calendar`'s time zone (default: the user's).
    init(date: Date, calendar: Calendar = .current) {
        let comps = Self.gregorian(zoneOf: calendar).dateComponents([.year, .month, .day], from: date)
        let utcMidnight = Self.utcCalendar.date(from: comps)!
        self.raw = Int((utcMidnight.timeIntervalSince1970 / Self.secondsPerDay).rounded(.down))
    }

    /// A specific Y/M/D (proleptic Gregorian) — mainly for tests and fixed dates.
    init(year: Int, month: Int, day: Int) {
        let utcMidnight = Self.utcCalendar.date(from: DateComponents(year: year, month: month, day: day))!
        self.raw = Int((utcMidnight.timeIntervalSince1970 / Self.secondsPerDay).rounded(.down))
    }

    /// TODAY is the training day, not the calendar day — see `rolloverHour`. Every
    /// "today" in the app comes through here (`DayClock`), so the tally, the strip, the
    /// grid and the day a session is filed under cannot disagree about when a day ends.
    /// `now` is a test seam.
    static func today(calendar: Calendar = .current, now: Date = .now) -> DayStamp {
        DayStamp(trainingDayOf: now, calendar: calendar)
    }

    // MARK: - The training day

    /// **A training day turns at 04:00, not at midnight.** A hang from 23:47 to just past
    /// midnight is an evening session; filing it under the morning after scored one
    /// evening as two days (2026-09-20). Anything before this hour counts for the day
    /// before: four is after any late session and before any morning one.
    static let rolloverHour = 4

    /// The training day `date` falls in: its calendar day, or the previous one when the
    /// local clock reads earlier than `rolloverHour`. Calendar arithmetic, never "minus
    /// four hours": on the night the clocks go forward, 04:30 minus four hours is 23:30
    /// the evening before.
    init(trainingDayOf date: Date, calendar: Calendar = .current) {
        let comps = Self.gregorian(zoneOf: calendar)
            .dateComponents([.year, .month, .day, .hour], from: date)
        var day = DayStamp(year: comps.year!, month: comps.month!, day: comps.day!)
        if let hour = comps.hour, hour < Self.rolloverHour { day = day - 1 }
        self = day
    }

    /// The instant the training day after `date`'s begins — the coming 04:00 local. What
    /// `DayClock` sleeps until, since UIKit posts nothing at that hour.
    static func nextRollover(after date: Date, calendar: Calendar = .current) -> Date {
        let rollover = DateComponents(hour: rolloverHour, minute: 0, second: 0)
        return Self.gregorian(zoneOf: calendar)
            .nextDate(after: date, matching: rollover, matchingPolicy: .nextTime)
            ?? date.addingTimeInterval(Self.secondsPerDay)
    }

    /// Local start-of-day for this day — for date pickers and chart axes. (If local
    /// midnight does not exist on a DST edge, Calendar picks the first valid instant.)
    func date(calendar: Calendar = .current) -> Date {
        let utcDate = Date(timeIntervalSince1970: Double(raw) * Self.secondsPerDay)
        let comps = Self.utcCalendar.dateComponents([.year, .month, .day], from: utcDate)
        return Self.gregorian(zoneOf: calendar).date(from: comps)!
    }

    /// "21 Jul", through the user's locale — never hand-assembled from month names.
    func formatted(calendar: Calendar = .current) -> String {
        date(calendar: calendar).formatted(.dateTime.day().month(.abbreviated))
    }
}

File: android/engine/src/main/kotlin/run/nuri/getagrip/engine/DayStamp.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.engine

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZonedDateTime

/// A calendar day as an integer — days since 1970-01-01 (proleptic Gregorian).
///
/// A training day is the day the user lived through, which has no timezone: a session
/// finished at 00:30 belongs to the evening it was part of — the day turns at
/// `ROLLOVER_HOUR`, not midnight — and "2 of 2 today" must flip at that hour without a
/// relaunch. Storing days as integers makes all of
/// that pure integer math and immune to the DST/timezone off-by-one bugs a timestamp
/// invites, and it makes `WorkoutLog.dayKey` a cheap Int predicate.
///
/// The ONLY place a time zone appears is this file's conversion boundary.
///
/// TRANSLATION NOTE (from Shared/Engine/DayStamp.swift): Swift takes the ZONE of the
/// user's `Calendar` and never its calendar system, because a Thai or Japanese device
/// calendar would shift every stamp by centuries. `java.time.LocalDate` is ISO-8601
/// (proleptic Gregorian) by construction and takes only a `ZoneId`, so the same rule
/// holds here without a special case — `LocalDate.toEpochDay()` IS `raw`.
data class DayStamp(val raw: Int) : Comparable<DayStamp>, JsonEncodable {

    override fun compareTo(other: DayStamp): Int = raw.compareTo(other.raw)

    // Strideable — the 14-day consistency strip is literally `(today - 13)..today`.
    fun advanced(by: Int): DayStamp = DayStamp(raw + by)
    fun distance(to: DayStamp): Int = to.raw - raw

    operator fun plus(days: Int): DayStamp = DayStamp(raw + days)
    operator fun minus(days: Int): DayStamp = DayStamp(raw - days)
    /// Signed distance in days (NOT a count of inclusive days — that is this + 1).
    operator fun minus(other: DayStamp): Int = raw - other.raw

    /// A bare Int on the wire — `20669`, never `{"raw":20669}`.
    override fun toJson(): JsonElement = JsonPrimitive(raw)

    fun localDate(): LocalDate = LocalDate.ofEpochDay(raw.toLong())

    /// Local start-of-day for this day — for date pickers and chart axes. If local
    /// midnight does not exist on a DST edge, `java.time` picks the first valid instant,
    /// exactly as Foundation's `Calendar` does.
    fun startOfDay(zone: ZoneId = ZoneId.systemDefault()): ZonedDateTime =
        localDate().atStartOfDay(zone)

    companion object {
        fun fromJson(e: JsonElement?): DayStamp? = JsonRead.int(e)?.let { DayStamp(it) }

        /// A specific Y/M/D (proleptic Gregorian) — mainly for tests and fixed dates.
        fun of(year: Int, month: Int, day: Int): DayStamp =
            DayStamp(LocalDate.of(year, month, day).toEpochDay().toInt())

        /// The calendar day `instant` falls on in `zone` (default: the user's).
        fun of(instant: Instant, zone: ZoneId = ZoneId.systemDefault()): DayStamp =
            DayStamp(instant.atZone(zone).toLocalDate().toEpochDay().toInt())

        /// TODAY is the training day, not the calendar day — see `ROLLOVER_HOUR`. Every
        /// "today" in the app comes through here (`DayClock`), so the tally, the strip, the
        /// grid and the day a session is filed under cannot disagree about when a day ends.
        /// `now` is a test seam.
        fun today(zone: ZoneId = ZoneId.systemDefault(), now: Instant = Instant.now()): DayStamp =
            trainingDayOf(now, zone)

        // MARK: - The training day

        /// **A training day turns at 04:00, not at midnight.** A hang that starts at 23:47
        /// and ends 44 seconds past midnight is an evening session; filing it under the
        /// morning after scores one evening as two days, which is exactly what Nuri's own
        /// history showed on iOS (2026-09-20). The promise this file always made — a 00:30
        /// session belongs to the day the climber lived through — was never implemented
        /// until this constant existed: the clock simply turned at midnight. Anything in the
        /// small hours before this counts for the day before.
        const val ROLLOVER_HOUR = 4

        /// The training day `instant` falls in: its calendar day, or the previous one when
        /// the local clock reads earlier than `ROLLOVER_HOUR`. Calendar arithmetic, never
        /// "minus four hours": on the night the clocks go forward, 04:30 minus four hours
        /// is 23:30 the evening before.
        fun trainingDayOf(instant: Instant, zone: ZoneId = ZoneId.systemDefault()): DayStamp {
            val local = instant.atZone(zone)
            val date = if (local.hour < ROLLOVER_HOUR) local.toLocalDate().minusDays(1) else local.toLocalDate()
            return DayStamp(date.toEpochDay().toInt())
        }

        /// The instant the training day after `instant`'s begins — the coming 04:00 local.
        /// What `DayClock` sleeps until, since no system broadcast marks that hour.
        fun nextRollover(after: Instant, zone: ZoneId = ZoneId.systemDefault()): Instant {
            val local = after.atZone(zone)
            var candidate = local.toLocalDate().atTime(ROLLOVER_HOUR, 0).atZone(zone)
            if (!candidate.isAfter(local)) {
                candidate = local.toLocalDate().plusDays(1).atTime(ROLLOVER_HOUR, 0).atZone(zone)
            }
            return candidate.toInstant()
        }

        /// The range of days `from..to`, inclusive, in order.
        fun span(from: DayStamp, to: DayStamp): List<DayStamp> =
            if (to.raw < from.raw) emptyList() else (from.raw..to.raw).map(::DayStamp)
    }
}
